Add parse_ants function to UVData object

jburba opened the pull request:
Might need to be modified for pep8, but 100% of parse_ants covered by tests.

adampbeardsley commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -1548,3 +1548,160 @@ def test_get_feedpols():
+    nt.assert_is_instance(polarizations,type(None))
+
+    # Multiples baselines with polarizations as list
+    ant_str = '1r_2l,1l_3l,1r_11r'
```
We should make it clear in the documentation that if you pass the result from a selection like this, you don't just get the three baselines, but rather you get all the pairs of antennas with all the polarizations. In this case you end up with:
1r_2l, 1l_2l, 1r_2r, 1r_3l, 1l_3l, 1r_3r, 1r_11l, 1l_11l, and 1r_11r.
If the intent is to pass to select, this is the right thing to do, but it should be clear.

adampbeardsley commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -1548,3 +1548,160 @@ def test_get_feedpols():
+    nt.assert_items_equal(ant_pairs_nums,ant_pairs_data)
+    nt.assert_items_equal(polarizations,polarizations_data)
+
+    # Antenna numbers with - for avoidance
```
Oh I think I see later that it excludes that antenna, rather than include it. Is there a more practical way to test this? The examples given aren't all that useful because e.g. `'1_10'` gives you the same result as `'1_(-3,10)'`, so why would you use the latter? Is the minus useful when otherwise using wildcards? For example, all baselines with antenna 1, excluding baselines with antenna 3?
Hmm.. after playing myself a little, it looks like `'1,-3'` is equivalent to `'1,3'`. Is this correct behavior?

jburba replied:
The behavior of the minus sign now works "correctly."  If you pass '1,-1_3', all baselines including antenna 1 are kept except (1,3).  Similarly, if you pass '1,2,-3' to select via ant_str, all baselines including antennas 1 and 2 are kept except the pairs (1,3) and (2,3).  Corrected in `8d6c986`

adampbeardsley commented:
re: @bhazelton I think it would be useful for select to take a string like this, but get_* functions won't really synergize well with this format. The get functions only allow certain slices of the data - single baseline/pol, single baseline with all pols, or single pol with all baselines. It's possible to use the output of this function to sequentially call get_*, but I think that's user level. Might be a good example for the tutorial though.
I also wouldn't want to lose the current select interface. Maybe we can add a new keyword which, if set, just runs select using the output of parse_ants.

jburba commented:
@adampbeardsley I agree.  Taking a look at the get_* functions, the input syntax is doable from parse_ants output, but I designed the parser with the select function in mind.  Currently working on having parse_ants called from the select function via an antenna string.

jburba commented on pyuvdata/uvdata.py:
```diff
@@ -1774,3 +1775,147 @@ def antpairpol_iter(self, squeeze='default'):
+
+        Args:
+            ant_str: String containing antenna information to pass to select function.
+                Can be all, auto, cross, or combinations of antenna numers and polarization
```
Corrected

jburba commented on pyuvdata/uvdata.py:
```diff
@@ -1774,3 +1775,147 @@ def antpairpol_iter(self, squeeze='default'):
+        """
+
+        ant_re = r'(\(((-?\d+[lrxy]?,?)+)\)|-?\d+[lrxy]?)'
+        bl_re = '(^(%s_%s|%s),?)' % (ant_re, ant_re, ant_re)
```
This came straight from `aipy/scripting.py` but I did add support for circular polarizations.

jburba commented on pyuvdata/uvdata.py:
```diff
@@ -1774,3 +1775,147 @@ def antpairpol_iter(self, squeeze='default'):
+                                if not ant_tuple in ant_pairs_nums:
+                                    ant_pairs_nums.append(ant_tuple)
+                elif ant_str[str_pos:].upper().startswith('I'):
+                    polarizations.append(1)
```
Corrected

jburba commented on pyuvdata/uvdata.py:
```diff
@@ -1774,3 +1775,147 @@ def antpairpol_iter(self, squeeze='default'):
+                        ant_tuple = tuple((ant,ant))
+                        ant_pairs_nums.append(ant_tuple)
+                elif ant_str[str_pos:].startswith('cross'):
+                    for ant1 in ant_nums:
```
This goes back to the discussion from Slack regarding how much the parser should know about the object.  I previously had implemented a check that an antenna pair would not be output in `ant_pairs_nums` unless it was present in `self.get_antpairs()`.  Should I put this check back in?  Or should that check be left to the select function?

jburba replied:
Corrected in `a97f1af`

jburba commented on pyuvdata/uvdata.py:
```diff
@@ -1774,3 +1775,147 @@ def antpairpol_iter(self, squeeze='default'):
+                    pass
+                elif ant_str[str_pos:].startswith('auto'):
+                    for ant in ant_nums:
+                        ant_tuple = tuple((ant,ant))
```
Corrected

jburba commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -1548,3 +1548,160 @@ def test_get_feedpols():
+    ant_str = 'cross'
+    ant_pairs_nums,polarizations = uv.parse_ants(ant_str)
+    for ant_pair in ant_pairs_nums:
+        nt.assert_not_equal(ant_pair[0],ant_pair[1])
```
Corrected in `860ff6e`

bhazelton commented on docs/tutorial.rst:
```diff
@@ -237,7 +237,110 @@ c) Select a few antenna pairs to keep
+Can be passed lowercase or uppercase
+
+- i,I: keeps only Stokes I
+- q,V: keeps both Stokes Q and V
```
this needs an example.

jburba replied:
Is there a file distributed with pyuvdata that contains Stokes params?  The file I used for the other examples only has circular pols.

bhazelton replied:
No, there is not a file that is circulated with Stokes parameters, but for a test you can always change the values in the polarization array after reading in any of the test files and before testing the select.

bhazelton commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -1548,3 +1548,366 @@ def test_get_feedpols():
+    # Auto correlations
+    ant_str = 'auto'
+    nt.assert_raises(ValueError, uv.select, ant_str=ant_str, inplace=inplace)
+    # No auto correlations in this data
```
You also need a test with a data set that has auto correlations so you can test that you get the expected number back. We have an autocorrelation test somewhere, so we have at least one test data set with autocorrelations.

bhazelton commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -1548,3 +1548,366 @@ def test_get_feedpols():
+    uv2 = uv.select(ant_str=ant_str, inplace=inplace)
+    nt.assert_items_equal(uv2.get_antpairs(), uv.get_antpairs())
+    nt.assert_items_equal(uv2.get_pols(), uv.get_pols())
+    # All baselines in data are cross correlations
```
You also need a test with a data set that has auto correlations so you can test that this doesn't just always return everything.

bhazelton commented on docs/tutorial.rst:
```diff
@@ -1,4 +1,4 @@
-Tutorial
+polarizationsTutorial
```
Did you really want to change this?

bhazelton commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -1548,3 +1548,366 @@ def test_get_feedpols():
+    uv2 = uv.select(ant_str=ant_str, ant_pairs_nums=ant_pairs_nums,
+                            polarizations=polarizations, inplace=inplace)
+    nt.assert_items_equal(uv2.get_antpairs(), ant_pairs)
+    nt.assert_items_equal(uv2.get_pols(), pols)
```
This last test is great! You should probably also have test for the case where ant_str is set along with antenna_numbers or antenna_names.

jpober commented:
A couple of general comments/questions for everyone:

(1) It seems like we have some duplicate functionality in that an ant_str can be used to accomplish a lot of the same things that would be accomplished by e.g. antenna_nums or ant_pairs_nums.  I understand not wanting to remove existing functionality, but I think we should consider having select() only accept an ant_str and making sure all the existing selects can be performed with an ant_str (or, maybe, an antpol_str if we want to incorporate polarization functionality completely into the string).  I think it's too confusing to have some things doable only with ant_str and others only doable with other select arguments.

(2) A couple pieces of behavior that don't seem right; some are due to select() and not parse_ants, but the two really should work in tandem.

a) On a data file with only xx polarization visibilities, uv2 = uv.select(ant_str='10x',inplace=False) returns an error that polarization -7 (i.e. XY) isn't in the data.  I think this should be a valid select and return all baselines containing the x polarization of antenna 10.  Not sure whether the right place for a change is in select or parse_ants.

b) parse_ants doesn't handle string arguments ('all', 'cross', 'auto') plus polarizations (or at least I couldn't figure out how to do it.  This would be for a case where e.g. I wanted all autocorrelations in a specific polarization.  I suppose the polarizations argument in select can handle this, but see comment (1) above. 

c) I couldn't find support for getting e.g. all baselines containing antenna 10 in the xx polarization (e.g. '10x_x' or something like that).  

Overall, this is _very_ nice.  Good work!

jburba commented:
Added some functionality for the plotting script/notebook I am working on now with @jpober as a "replacement" for plot_uv.py from aipy which is why the coverage decreased.  The output formatting for this plotting functionality is not set in stone yet, so I figured there is no immediate reason to test it.  That bit of the code for plotting aside, there's 100% coverage of the code in the parse_ants function written specifically for the select function which is the main contribution of this pull request.

jburba commented:
Last clarifying comment hopefully.  I moved all of the code I added to parse_ants to another branch parse_ants_plotting to prevent any confusion when looking at the parse_ants function and its tests.  This branch now only contains the pieces of parse_ants that are designed to work with the select function.

bhazelton commented:
@jburba Can you please rebase your branch on master? There has been a significant change in the way we format the tutorial to support automatic documentation testing via Travis (to ensure that the tutorial does not get out of date). If you'd like help incorporating these changes into your PR after you rebase, I'd be happy to help.

bhazelton commented:
@jburba this looks like it's really close. It looks to me like there are a few outstanding comments that we should address. Some of them are mine about testing changes -- those should be pretty easy. 

@jpober's comments are more difficult to address. I think some of them are things that need to be discussed and decided by a small group of people while you may be able to address the others. We may need to have a short telecon about this with @adampbeardsley @dannyjacobs in order to make the decisions.

jpober commented:
I think my comments have largely been addressed, although I haven't looked at the code in a while.  I think the plan was to accept either the existing syntax or the ant_str in select() but to not accept both simultaneously and to provide clear erroring when both are provided.  @jburba can correct me if I'm wrong.

bhazelton commented:
@jburba I guess not. I was trying to remember where we were on this PR and It looked like there were review comments about testing that hadn't been addressed, but I looked more carefully at the code and it looks like you have actually addressed all of them.
